### PR TITLE
Install devDependencies by default

### LINKED ISSUE #431: Consider installing devDependencies by default
Pro: Many users getting started with Heroku try to run a build step with `webpack`, `gulp`, etc only to find that it's in their `devDependencies` and not installed by default on Heroku. They then have to figure out how to tell Heroku to install these. Defaulting to installing everything makes for a smoother onboarding.

Con: Longer install times and larger slugs for existing large apps

First step: gather metrics

### PR DESCRIPTION
**Pull request by jmorrell**
Will fix #431 by making the default behavior to: 

- install dependencies + devDependencies
- run the build scripts
- cache `node_modules` for the next build
- prune `devDependencies` from `node_modules`

If a user needs to customize this behavior, they should be able to:

#### Only install `dependencies`, the current default behavior

You can get this behavior by setting the following env vars:
`heroku config:set NPM_CONFIG_PRODUCTION=true YARN_PRODUCTION=true`

#### Install both `dependencies` and `devDependencies` but do not prune

You can get this behavior by setting the following env vars:
`heroku config:set NPM_CONFIG_PRODUCTION=false YARN_PRODUCTION=false`

#### `NODE_ENV != production`

If `NODE_ENV` is not set to `production` (ex: `test` or `staging`) then we will install `devDependencies` and skip pruning

### CODE REVIEW

**Comment by edmorley** [comment]
Would it be possible to provide an option to opt out of this? As-is this will mean we install/cache packages that we don't need.

It's such a shame that `package.json` doesn't have the concept of runtime vs buildtime vs devDependencies, since the user confusion (and thus the need for this PR) really stems from buildtime deps having to be shoe-horned into either `dependencies` or `devDependencies` when really neither is always correct (particularly for apps vs libraries).

**Comment by jmorrell** [comment]
> Would it be possible to provide an option to opt out of this? As-is this will mean we install/cache packages that we don't need.

@edmorley I intend to make it so that you can configure it to only install `dependencies` or install both and skip pruning. I'll make sure to document that.

I believe this will be a better default for most users though.

**Comment by hunterloftis** [comment]
Agreed that it's a shame there's no concept of build dependencies (for things like build, but also CI).

While this will install `devDependencies`, based on the prune step @jmorrell mentioned above, I don't think it will cache them, so at least they won't bloat tarballs or increase transfer times.

**Comment by jmorrell** [comment]
@hunterloftis I'm currently pruning after the cache is saved. Otherwise the user needs to re-download all of the build tooling again the next time.

**Comment by hunterloftis** [comment]
Ah yeah, I hadn't thought about that. Glad you have :)

File: bin/compile [hunk 1]
```diff
@@ -211,6 +211,17 @@ cache_build() {
...
 
+prune_devdependencies() {
+  if $YARN; then
+    yarn_prune_devdependencies "$BUILD_DIR"
+  else
+    npm_prune_devdependencies "$BUILD_DIR"
+  fi
+}
+
+header "Pruning devDependencies"
```
Comment (jmorrell) [inline bin/compile]: This is the terminology that npm uses, but I don't love it.

Is there a clearer way to get across "we're removing the modules defined in `devDepenencies` from `node_modules`"?

**Comment by jmorrell** [comment]
@edmorley I can't tag you as a reviewer, but I'd appreciate it if you took a look :)

File: bin/compile [hunk 2]
```diff
@@ -211,6 +211,17 @@ cache_build() {
 header "Caching build"
```
Comment (jmorrell) [inline bin/compile]: This changes a lot of the fixture files for tests which can be mostly ignored. The important changes are found in:

`bin/compile`
`lib/dependencies.sh`
`lib/environment.sh`
`lib/failure.sh`

With tests changes in:
`test/run`

which sadly gets shuffled to the bottom in this github view :(

File: lib/failure.sh [hunk 3]
```diff
@@ -122,8 +122,7 @@ fail_yarn_outdated() {
   local log_file="$1"
   local yarn_engine=$(read_json "$BUILD_DIR/package.json" ".engines.yarn")
 
-  if grep -qi 'error: unknown option .--frozen-lockfile' "$log_file"; then
-    echo "ran"
+  if grep -qi 'error .install. has been replaced with .add. to add new dependencies' "$log_file"; then
```
Comment (jmorrell) [inline lib/failure.sh]: This is changed because in these earlier versions adding the `--production` flag changes the error message that is displayed

File: test/run [hunk 4]
```diff
@@ -641,13 +644,6 @@ testEnvVars() {
   assertCapturedSuccess
 }
 
-testNoEnvVars() {
-  env_dir=$(mktmpdir)
-  compile "stable-node" "$(mktmpdir)" $env_dir
-  assertCaptured "NPM_CONFIG_PRODUCTION=true"
```
Comment (jmorrell) [inline test/run]: We're no longer setting `NPM_CONFIG_PRODUCTION` by default

File: test/run [hunk 5]
```diff
@@ -263,10 +270,6 @@ testWarnDevDeps() {
   assertCaptured "A module may be missing"
   assertNotCaptured "This module may be specified"
   assertCapturedError
-  compile "missing-devdeps-2"
```
Comment (jmorrell) [inline test/run]: This will no longer result in an error so I've removed it

File: test/run [hunk 6]
```diff
@@ -712,7 +777,7 @@ testShrinkwrap() {
   compile "shrinkwrap"
   assertCaptured "express@4.10.4"
   assertCaptured "lodash@2.4.0"
-  assertNotCaptured "mocha"
+  assertCaptured "mocha@2.0.1"
```
Comment (jmorrell) [inline test/run]: This is now installed by default

File: test/run [hunk 7]
```diff
@@ -811,6 +918,10 @@ compileTest() {
   for f in ${compile_dir}/.profile.d/*; do source $f > /dev/null 2> /dev/null ; done
 
   capture ${bp_dir}/bin/test ${compile_dir}
+
```
Comment (jmorrell) [inline test/run]: Fixes a bug in the `compileTest` function that would cause subsequent tests to fail

File: bin/compile (hunk 1 above)
Comment (hone) [reply bin/compile]: If it's the command in both `yarn` and `npm`, it's probably fine.

File: lib/dependencies.sh [hunk 8]
```diff
@@ -87,14 +87,34 @@ log_build_scripts() {
...
   echo "Installing node modules (yarn.lock)"
   cd "$build_dir"
-  yarn install --frozen-lockfile --ignore-engines 2>&1
+  yarn install --production=$production --frozen-lockfile --ignore-engines 2>&1
+}
+
+yarn_prune_devdependencies() {
+  local build_dir=${1:-} 
+
+  if [ $NODE_ENV == "test" ]; then
```
Comment (hone) [inline lib/dependencies.sh]: This will be caught by the below. Do we want the explicit messaging here?

File: lib/dependencies.sh [hunk 9]
```diff
@@ -124,8 +145,32 @@ npm_rebuild() {
...
   else
     echo "Skipping (no package.json)"
   fi
 }
+
+npm_prune_devdependencies() {
+  local build_dir=${1:-} 
+  local npm_version=$(npm --version)
+
+  if [ $NODE_ENV == "test" ]; then
```
Comment (hone) [inline lib/dependencies.sh]: Same as the yarn comment above, but wanted to point it out for completeness sake.

**Comment by hone** [comment]
I ran through all the (non test) code. This looks good. I tried to find a OS X event library in node, but it looks like all file watcher libs are cross platform that are popular. I'm not sure of other platform specific libs that would only show up in node.

As a suggestion to provide helpful messaging to users who builds work now but may break with this change, the buildpack can mark in the cache the buildpack version number with this PR. If the build fails and the last build was not using this buildpack and the last build was not a success, then we can display how to opt into the old functionality.

File: lib/dependencies.sh [hunk 10]
```diff
@@ -87,14 +87,34 @@ log_build_scripts() {
...
+  yarn install --production=$production --frozen-lockfile --ignore-engines 2>&1
+}
+
+yarn_prune_devdependencies() {
+  local build_dir=${1:-} 
+
+  if [ $NODE_ENV == "test" ]; then
+    echo "Skipping because NODE_ENV is 'test'"
+    return 0
+  elif [ $NODE_ENV != "production" ]; then
```
Comment (edmorley) [inline lib/dependencies.sh]: It might be worth quoting the `$NODE_ENV` and the various other variables in conditionals, since otherwise spurious spaces can cause a `too many arguments` bash error and the wrong result.

eg:
```bash
$ NODE_ENV="foo bar"; if [ $NODE_ENV != "production" ]; then echo "not production"; else echo "production"; fi
bash: [: too many arguments
production
```

Fwiw [shellcheck](https://github.com/koalaman/shellcheck) is able to find issues like this (the Python buildpack runs it on Travis):
```bash
In lib/dependencies.sh line 103:
  elif [ $NODE_ENV != "production" ]; then
         ^-- SC2086: Double quote to prevent globbing and word splitting.
```
(https://github.com/koalaman/shellcheck/wiki/SC2086)

Comment (kennethreitz) [reply lib/dependencies.sh]: *cough* shellcheck is the best

**Comment by Chronial** [comment]
Just a heads up: The silent removal of `NPM_CONFIG_PRODUCTION` silently broke our deploy process. We only actually build our code on `npm install` if that variable was set, so a version without any js code was deployed.

**Comment by jmorrell** [comment]
@Chronial I'm sorry to hear that 😓 Thank you for letting me know. I've tried to be as backward-compatible as I compatible as I could, but this is likely a breaking change for a small subset of users.

If you need to test for an env var, `NODE_ENV=production` is probably the one you want, or better, put the build into a `heroku-postbuild` script: https://devcenter.heroku.com/articles/nodejs-support#heroku-specific-build-steps

**Comment by Chronial** [comment]
> put the build into a heroku-postbuild script: https://devcenter.heroku.com/articles/nodejs-support#heroku-specific-build-steps

Thanks, that's what we did. I think Heroku would benefit from a mechanism to notify its users about breaking changes like this before they happen.

**Comment by tomreid76** [comment]
This also broke our deploys because we use private devDependencies. A heads up to paying customers would have been really nice instead of sending us into a panic and having to update environment variables across many applications.

**Comment by Cryptophobia** [comment]
Apart from the breaking changes, thank you for this much needed change to the default behaviour of the buildpack!

**Comment by jmorrell** [comment]
@Chronial @tomreid76 I don't disagree at all and could have done a better job here. It's a difficult balance because sending out mass emails can cause more stress and confusion than it alleviates. We saw this when sending out security notices and getting hundreds of support tickets from confused users.

Thanks for the feedback, and we'll review this to figure out how we can improve messaging next time.

**Comment by jmorrell** [comment]
@tomreid76 Can you elaborate on how private devDependencies were effected?

**Comment by tomreid76** [comment]
Sure. We have a few private GitHub repos as devDependencies. Before, they were not installed by default when deploying (using node 6.11.x and npm 3.10.10). This is exactly how we wanted it. As of the update yesterday, Heroku started attempting install of those respective devDeps and things started blowing up because Heroku has no access.

One other thing to note is that we actually have a pre-deploy step wherein we generate a versioned build artifact in Travis CI, and so our deploy to Heroku is really just a deploy since everything needed (prod dependencies) is bundled.

Admittedly perhaps not the most common use case, but a "verified" artifact is important for reasons I won't get into. :)

In any case, problem is solved, and so not a big deal, just in the future would have appreciate an email or something... I understand that has consequences also though.

This also looks like a significant undertaking and I don't want to downplay those efforts or anything. Thanks for all your efforts in what up to yesterday has been a fairly good and seamless experience for me and others on my team.

**Comment by edmorley** [comment]
Perhaps for more significant changes a "change X is planned for release in N days" entry to https://devcenter.heroku.com/changelog (in addition to the one made post release) would be visible but yet not as spammy as an email? (Anyone who doesn't want to be surprised by changes should subscribe to that feed by RSS IMO)

**Comment by Chronial** [comment]
@jmorrell a suggestion for a simple notification system: Open an issue in the github repo titled "Breaking Changes Notifications" and just add a comment there a few days before you deploy breaking changes. Everybody who is interested can just subscribe to that issue. Since it's opt-in this should not cause any confusion.

Then maybe reference that issue in this Repo's README and the heroku node.js docs.

**Comment by darrenklein** [comment]
@jmorrell I set the following env variables on my app

`heroku config:set NPM_CONFIG_PRODUCTION=true YARN_PRODUCTION=true NODE_ENV=production`

but Heroku is still trying to install my `devDependencies` which is breaking my deployment. What might I be missing?

**Comment by jmorrell** [comment]
@darrenklein That doesn't sound right. Can you open a support ticket at help.heroku.com and give me the ticket #? I'm happy to take a look

**Comment by jjalan** [comment]
@jmorrell - I have set up for my app

heroku config:set NPM_CONFIG_PRODUCTION=true YARN_PRODUCTION=true NODE_ENV=production

However, review apps created do not inherit these settings.

Is this a bug or design? If by design, how can we enable set NPM_CONFIG_PRODUCTION=true YARN_PRODUCTION=true for review apps?

**Comment by jjalan** [comment]
@jmorrell - I guess one way I could think about that is setting them to true in app.json env section.

**Comment by darrenklein** [comment]
@jmorrell I actually have a ticket open, which is how I was directed to this thread... it's ticket #560469.

Here's the error log from the breaking deployment - you can see that those env variables have been set, but the package that it's breaking on is from my `devDependencies`.

```
remote: Compressing source files... done.
remote: Building source:
remote: 
remote: -----> Node.js app detected
remote: 
remote: -----> Creating runtime environment
remote:        
remote:        NPM_CONFIG_LOGLEVEL=error
remote:        NPM_CONFIG_PRODUCTION=true
remote:        YARN_PRODUCTION=true
remote:        NODE_VERBOSE=false
remote:        NODE_ENV=production
remote:        NODE_MODULES_CACHE=true
remote: 
remote: -----> Installing binaries
remote:        engines.node (package.json):  8.9.4
remote:        engines.npm (package.json):   5.7.1
remote:        
remote:        Resolving node version 8.9.4...
remote:        Downloading and installing node 8.9.4...
remote:        Bootstrapping npm 5.7.1 (replacing 5.6.0)...
remote:        npm 5.7.1 installed
remote: 
remote: -----> Restoring cache
remote:        Skipping cache restore (new-signature)
remote: 
remote: -----> Building dependencies
remote:        Installing node modules (package.json + package-lock)
remote:        npm ERR! Error while executing:
remote:        npm ERR! /usr/bin/git ls-remote -h -t ssh://git@gitlab.com/my-team/eslint-config-custom-config.git
remote:        npm ERR!
remote:        npm ERR! Host key verification failed.
remote:        npm ERR! fatal: Could not read from remote repository.
remote:        npm ERR!
remote:        npm ERR! Please make sure you have the correct access rights
remote:        npm ERR! and the repository exists.
remote:        npm ERR!
remote:        npm ERR! exited with error code: 128
remote:        
remote:        npm ERR! A complete log of this run can be found in:
remote:        npm ERR!     /tmp/npmcache.hhLmL/_logs/2018-03-07T23_09_52_031Z-debug.log
remote: 
remote: -----> Build failed
remote:        
remote:        We're sorry this build is failing! You can troubleshoot common issues here:
remote:        https://devcenter.heroku.com/articles/troubleshooting-node-deploys
remote:        
remote:        If you're stuck, please submit a ticket so we can help:
remote:        https://help.heroku.com/
remote:        
remote:        Love,
remote:        Heroku
remote:        
remote:  !     Push rejected, failed to compile Node.js app.
remote: 
remote:  !     Push failed
remote: Verifying deploy...
remote: 
remote: !	Push rejected to my-app-12345.
remote: 
```

From the `package-lock.json`:
```
    "eslint-config-custom-config": {
      "version": "git+ssh://git@gitlab.com/my-team/eslint-config-custom-config.git#859a39ac261140d442919f0f36dae5e0a2cf322a",
      "dev": true
    },
```

I also tried setting the buildpack to an earlier version, but it was still resulting in the same failure, which makes me think that something else may be up. Any help you can offer would be much appreciated.

**Comment by mars** [comment]
@darrenklein [said](https://github.com/heroku/heroku-buildpack-nodejs/pull/519#issuecomment-371539305),

```
remote:        npm ERR! Error while executing:
remote:        npm ERR! /usr/bin/git ls-remote -h -t ssh://git@gitlab.com/my-team/eslint-config-custom-config.git
remote:        npm ERR!
remote:        npm ERR! Host key verification failed.
remote:        npm ERR! fatal: Could not read from remote repository.
```

How the Gitlab ssh authentication configured? Where is the private key? Does it come from another buildpack? If this is a Review App, is it missing a config var for auth?

**Comment by jmorrell** [comment]
@mars I'm answering him in a support ticket :)

tl;dr - When you have a private git repo in your `package-lock.json`, npm tries to resolve it when building out the dependency tree, even if it won't be installed. This would have been an issue before this change as well.

I'm not sure if this is a bug in npm or a limitation of its algorithm.

In this case it seems like moving the git dependency to `optionalDependencies` fixes the build without disrupting the local workflow.
